torch/csrc/jit/passes/utils/subgraph_matcher.cpp

```diff
@@ -0,0 +1,277 @@
+#include <torch/csrc/jit/passes/utils/subgraph_matcher.h>
+
+namespace torch {
+namespace jit {
+
+MatchIterator& MatchIterator::begin() {
+  tried_nodes_.clear();
+  updateMatch();
+  return *this;
+}
+
+MatchIterator& MatchIterator::end() {
+  return *this;
+}
+
+MatchIterator& MatchIterator::operator++() {
+  updateMatch();
+  return *this;
+}
+
+MatchIterator& MatchIterator::operator++(int) {
+  updateMatch();
+  return *this;
+};
+
+MatchIterator::Match MatchIterator::operator*() {
+  return match_;
+}
+
+bool MatchIterator::operator!=(MatchIterator& rhs) {
+  if (finished_) {
+    return false;
+  }
+  return true;
+}
+
+const Node* MatchIterator::findNewNode() {
+  for (const auto node : graph_.nodes()) {
```

Comment:
Why not just use a node iterator to track which nodes have been visited (`graph_.nodes().begin()`)?

Reply:
does this iterator work if the graph is mutated?